src/printer.js

```diff
@@ -609,24 +605,45 @@ function genericPrintNoParens(path, options, print) {
     case "ReturnStatement":
       parts.push("return");
 
-      if (
-        n.argument &&
-        n.argument.comments &&
-        n.argument.comments.some(comment => comment.leading)
-      ) {
-        parts.push(
-          concat([
-            " (",
-            indent(
-              options.tabWidth,
-              concat([softline, path.call(print, "argument")])
-            ),
-            line,
-            ")"
-          ])
-        );
-      } else if (n.argument) {
-        parts.push(" ", path.call(print, "argument"));
+      if (n.argument) {
+        const comments = n.argument.comments || [];
+        if (
+          n.argument.type === "BinaryExpression" ||
+          n.argument.type === "LogicalExpression" ||
+          n.argument.type === "ConditionalExpression" ||
+          n.argument.type === "CallExpression" ||
+          n.argument.type === "MemberExpression" ||
+          n.argument.type === "SequenceExpression" ||
+          n.argument.type === "TaggedTemplateExpression"
+        ) {
+          function getLeftSide(node) {
+            if (node.expressions) return node.expressions[0];
+            return node.left || node.test || node.callee || node.object || node.tag;
+          }
+
+          let leftMost = n.argument;
+          let newLeftMost;
+          while (newLeftMost = getLeftSide(leftMost)) {
+            leftMost = newLeftMost;
+            if (leftMost.comments) comments.push(...leftMost.comments);
+          }
+        }
+
+        if (comments && comments.some(comment => comment.leading)) {
```

Comment:
Also, comment.leading is too broad, I later implemented `hasLeadingOwnLineComment` which is a more restrictive one. For example, it doesn't match 

```js
/* comment */ node
```

which wouldn't be an issue here.